### PR TITLE
[hist] THnSparse::Scale iterate only over non-filled bins

### PR DESCRIPTION
**Pull request by ferdymercury**
# This Pull request:

## Changes or fixes:

Fixes https://github.com/root-project/root/issues/13962

## Checklist:

- [ ] tested changes locally
- [x] updated the docs (if necessary)

This PR fixes #

### CODE REVIEW

File: hist/hist/src/THnSparse.cxx [hunk 1]
```diff
@@ -942,6 +942,31 @@ void THnSparse::AddBinError2(Long64_t bin, Double_t e2)
...
+/// It does not modify the histogram's number of entries.
+
+void THnSparse::Scale(Double_t c)
+{
+
+   Double_t nEntries = GetEntries();
+   // Scale the contents & errors
+   Bool_t haveErrors = GetCalculateErrors();
+   Long64_t i = 0;
+   THnSparseBinIter iter(kFALSE, this);
```
Comment (guitargeek) [inline hist/hist/src/THnSparse.cxx:957]:
```suggestion
   std::unique_ptr<ROOT::Internal::THnBaseBinIter> iter{CreateIter(false)};
```
There is already a virtual method to create the right iterator. Can we just use that one in `THnBase::Scale()`? Then you don't need to make `Scale()` virtual.

Also, there are many occurrences on explicit `THnIter` instantiations in `THnBase`, e.g. also in `Multiply()`. Can we replace all of them with the virtual iterator creation dispatch in one go?

**Review by guitargeek — DISMISSED** [review]
Thanks a lot for the initiative! I have a suggestion to approach the problem in a more general way.

File: hist/hist/src/THnSparse.cxx (hunk 1 above)
Comment (ferdymercury) [reply hist/hist/src/THnSparse.cxx]: Done! thanks for the suggestion

**Comment by dpiparo** [comment]
@guitargeek : can we merge to then backport to the 6.32 branch?

**Review by guitargeek — APPROVED** [review]
Awesome, thanks for following up!

**Pull request merged**